tests/manager/sokovan/deployment/handlers/test_handler_lock_concurrency.py

```diff
@@ -0,0 +1,141 @@
+            if concurrent_execution_flag["is_executing"]:
+                execution_log.append(("concurrent_detected", time.time(), call_id))
+
+            concurrent_execution_flag["is_executing"] = True
+            await asyncio.sleep(0.1)  # Simulate work
+            concurrent_execution_flag["is_executing"] = False
+
+            return MagicMock(successes=[], errors=[])
+
+        mock_executor = MagicMock()
+        mock_executor.provision_routes = AsyncMock(side_effect=mock_provision_routes)
+        return mock_executor
+
+    @pytest.fixture
+    def provisioning_handler(self, mock_route_executor: MagicMock) -> ProvisioningRouteHandler:
+        """Real ProvisioningRouteHandler instance with mocked executor"""
+        mock_event_producer = MagicMock()
+        return ProvisioningRouteHandler(mock_route_executor, mock_event_producer)
+
+    @pytest.fixture
+    async def route_coordinator(
+        self, mock_lock_context: Any, provisioning_handler: ProvisioningRouteHandler
+    ) -> AsyncGenerator[RouteCoordinator, None]:
+        """Create RouteCoordinator with real ProvisioningRouteHandler"""
+        mock_lock_factory = MagicMock(side_effect=mock_lock_context)
+        mock_valkey = AsyncMock()
+        mock_deployment_repo = AsyncMock()
+        mock_deployment_repo.get_routes_by_statuses = AsyncMock(return_value=[])
+        mock_event_producer = MagicMock()
+        mock_config_provider = MagicMock()
+        mock_config_provider.config.manager.session_schedule_lock_lifetime = 1.0
+
+        coordinator = RouteCoordinator(
+            valkey_schedule=mock_valkey,
+            deployment_repository=mock_deployment_repo,
+            event_producer=mock_event_producer,
+            lock_factory=mock_lock_factory,
+            config_provider=mock_config_provider,
+            scheduling_controller=MagicMock(),
+            client_pool=MagicMock(),
+            service_discovery=MagicMock(),
+        )
+        coordinator._route_handlers = {RouteLifecycleType.PROVISIONING: provisioning_handler}
+
+        yield coordinator
+
+    @pytest.mark.asyncio
+    async def test_concurrent_execution_prevented_by_lock(
+        self,
+        route_coordinator: RouteCoordinator,
+        execution_log: list[tuple[str, float, int]],
+        lock_acquired_count: dict[str, int],
+    ) -> None:
+        """
+        Integration test: Verify that lock prevents concurrent execution.
+
+        This test simulates the scenario where short cycle and long cycle
+        try to execute the same handler concurrently. With lock_id, the second
+        execution should wait for the first one to complete.
+        """
+        # Execute two concurrent calls (simulating short cycle + long cycle)
+        await asyncio.gather(
+            route_coordinator.process_route_lifecycle(RouteLifecycleType.PROVISIONING),
+            route_coordinator.process_route_lifecycle(RouteLifecycleType.PROVISIONING),
```

Comment:
Not only this scenario, but we should also look at the "if process needed" scenario. For now, please remove the test.
This scenario seems prone to timing issues, even if it cannot actually run.